sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc

```diff
@@ -0,0 +1,270 @@
+== Issues
+
+* How should we expose the difference between OpenCL C versions?
+  It seems like there are two aspects to the problem.
+  Applications need some way to query which versions the backend (or device)
+  supports.
+  Applications also need some way to tell the runtime which version the kernel
+  is written in.
++
+--
+One option is to define separate enumerators in `source_language` for each
+version like this:
+
+```
+enum class source_language : /*unspecified*/ {
+  opencl_1_0,
+  opencl_1_1,
+  opencl_2_0,
+  opencl_3_0,
+};
+```
+
+Applications could then query the supported versions via
+`is_source_kernel_bundle_supported`, and applications would identify the
+version of their kernel string via the `lang` parameter to
+`create_kernel_bundle_from_source`.
+
+Alternatively, this extension could define just a single language enumerator
+(`opencl`), but also provide as separate query to get the supported OpenCL C
+versions.
+When building a kernel bundle, applications would be required to pass "-cl-std"
+via the `build_options` property in order to identify the OpenCL C version of
+their source string.
+--
+
+* How can an application determine the OpenCL C optional features that are
+  supported and the extensions that are supported?
+  One option is to require the application to use OpenCL APIs for these
+  queries.
+  This seems better than duplicating these queries into this extension.
+  However, this assumes the application is running with an OpenCL backend.
+  Do we want to support the use of OpenCL C kernels also with the Level Zero
+  backend?
+  Currently, the online_compiler does support this case (but it provides no way
+  to query about optional features or extensions).
+
+* There should be a more formal definition of the kernel arguments that can be
+  passed.
+  For example, can the application pass a `local_accessor` as an argument to an
+  OpenCL C kernel?
+  We should have a complete list of allowed arguments and describe how each is
+  passed from the SYCL host code.
+
+* Do we need to document some restrictions on the OpenCL C
+  https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#work-item-functions[
+  work-item functions] that the kernel can call, which depends on how the
+  kernel was launched?
+  For example, can a kernel launched with the simple `range` form of
+  `parallel_for` call `get_local_size`?
+  In OpenCL, there is only one way to launch kernels
+  (`clEnqueueNDRangeKernel`), so it is always legal to call any of the
+  work-item functions.
+  If an OpenCL kernel is launched with a NULL `local_work_size` (which is
+  roughly equivalent to SYCL's `range` form of `parallel_for`), the
+  `get_local_size` function returns the local work-group size that is chosen by
+  the implementation.
+  Level Zero, similarly, has only one way to launch kernels.
+  Therefore, maybe it is OK to let kernels in this extension call any of the
+  work-item functions, regardless of how they are launched?
```

Comment:
I'm still not sure what to do about this issue, so I'd like to address it also in a follow-on PR.  I think we decided against having a separate invocation function like `invoke_native_kernel` because we didn't add anything like that in #11491.  Therefore, I think we'll probably solve this through some documentation in the specification.

Reply:
Yes, I agree.  My comment here predated the discussion we had about the enqueue functions.

I think it does make sense to limit the language features that are available to basic `parallel_for`, though.  The specification allows basic and ND-range parallel for kernels to have different execution models, which means that some implementations (not OpenCL or Level Zero) might launch them completely differently.  Some catch-all statement about how kernels constructed via online compilation must still obey the execution model implied by the function used to launch them seems easier to specify than maintaining a list of weird corner cases.